### ISSUE TITLE
[BUG] Validation hangs during "Initializing for testing transcriptions"

### ISSUE THREAD

**Issue by DanielSWolf** [MontrealCorpusTools/Montreal-Forced-Aligner#395]
**Debugging checklist**

[x] Have you updated to latest MFA version? *2.0.0rc3*
[x] Have you tried rerunning the command with the `--clean` flag?

**Describe the issue**

My corpus has some errors in the transcriptions (see [this comment](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner/issues/386#issuecomment-1013777759)), so I ran validation with the `--test_transcriptions` flag in order to fix or remove the incorrect transcripts. The validation script quickly reached the output line "Initializing for testing transcriptions...", then stopped producing any further output and seemed to hang.

I let it run over night, but there was no further output, so I killed it. In contrast, validation without the `--test_transcriptions` flag takes less than half an hour on my machine.

**For Reproducing your issue**
Please fill out the following:

1. Corpus structure
   * What language is the corpus in? Italian
   * How many files/speakers? 62,133 sound files, 85 speakers.
   * Are you using lab files or TextGrid files for input? lab.
2. Dictionary
   * Are you using a dictionary from MFA? If so, which one? No, a custom Italian dictionary.
   * If it's a custom dictionary, what is the phoneset? It's IPA, but with non-trivial phonemes like "mː" and "lː" that MFA doesn't know about. So I didn't specify a phone set during training.

**Log file**

[validate_training.log](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner/files/7931260/validate_training.log) *Note:* The last line ("Done! ...") only appeared when I killed the process. The script didn't terminate on its own.

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Version 20.04.3 LTS
 - WSL on Windows 10

**Additional context**

In case it helps, here is the call stack that appeared when I killed the script:

```text
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/daniel/miniconda3/bin/mfa", line 11, in <module>
    sys.exit(main())
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/command_line/mfa.py", line 929, in main
    run_validate_corpus(args, unknown)
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/command_line/validate.py", line 99, in run_validate_corpus
    validate_corpus(args, unknown)
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/command_line/validate.py", line 45, in validate_corpus
    validator.validate()
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/validator.py", line 994, in validate
    self.setup()
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/validator.py", line 966, in setup
    self.initialize_utt_fsts()
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/validator.py", line 747, in initialize_utt_fsts
    self.output_utt_fsts()
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/validator.py", line 295, in output_utt_fsts
    utt2fst = self.utt2fst_scp_data(num_frequent_words)
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/validator.py", line 273, in utt2fst_scp_data
    lookup = utterance.speaker.dictionary.split_clitics(t)
  File "/home/daniel/miniconda3/lib/python3.9/site-packages/montreal_forced_aligner/dictionary/pronunciation.py", line 224, in split_clitics
    return self.construct_split_words_function()(item)
KeyboardInterrupt
```

**Comment by DanielSWolf** [comment]
It seems that MFA isn't actually hanging; it's just taking very long. I believe that there are two reasons.

1. `get_word_frequency` seems to be very slow. It's meant to calculate the frequency of every word in the corpus by iterating over the utterances. The problem is that on my machine, it takes about 170 ms per utterance. For 60,000 utterances, this means that this function alone takes about 3 hours to complete.
2. I'm not sure about this point, but the code in `utt2fst_scp_data` looks to me as if `get_word_frequency` might be called more than once during validation, even though it doesn't take any arguments (and thus probably does the exact same thing on each call).

To find out why `get_word_frequency` is so slow, I profiled a short stretch of time. Here's what I found:

* The *entire* time is spent in `split_clitics`. <br/> In my case, that's entirely unnecessary because I already normalized all utterances. There's nothing to split here.
* 67% of `split_clitics` is spent in `construct_split_word_function`. So two thirds of the processing time might be saved by reusing the constructed function.
